File: internal/api/system.go
package api

import (
	"fmt"
	"net"
	"net/http"
	"runtime"
	"strings"
	"time"

	"github.com/clash-proxyd/proxyd/internal/types"
	"github.com/gin-gonic/gin"
)

// SystemInfo represents system information
type SystemInfo struct {
	Version               string     `json:"version"`
	GoVersion             string     `json:"go_version"`
	Uptime                int64      `json:"uptime"`
	MihomoStatus          string     `json:"mihomo_status"`
	Database              string     `json:"database"`
	StartTime             time.Time  `json:"start_time"`
	RuntimeConfigPath     string     `json:"runtime_config_path"`
	LastAutoUpdateAction  string     `json:"last_auto_update_action,omitempty"`
	LastAutoUpdateDetails string     `json:"last_auto_update_details,omitempty"`
	LastAutoUpdateAt      *time.Time `json:"last_auto_update_at,omitempty"`
	LastAlertAction       string     `json:"last_alert_action,omitempty"`
	LastAlertDetails      string     `json:"last_alert_details,omitempty"`
	LastAlertAt           *time.Time `json:"last_alert_at,omitempty"`
}

var startTime = time.Now()

// GetSystemInfo returns system information
func (h *Handler) GetSystemInfo(c *gin.Context) {
	uptime := time.Since(startTime).Seconds()

	// Get mihomo status
	mihomoStatus := "stopped"
	if runtime, err := h.runtimeStore.Get(); err == nil {
		mihomoStatus = runtime.Status
	}

	// Use the configured default path — this is where new configs should be written.
	runtimeConfigPath := h.defaultRuntimeConfigPath()

	info := SystemInfo{
		Version:           "1.0.0",
		GoVersion:         runtime.Version(),
		Uptime:            int64(uptime),
		MihomoStatus:      mihomoStatus,
		Database:          "sqlite",
		StartTime:         startTime,
		RuntimeConfigPath: runtimeConfigPath,
	}

	hydrateLastAutoUpdateSummary(&info, h.auditStore)
	hydrateLastAlertSummary(&info, h.auditStore)
	h.respondJSON(c, http.StatusOK, info)
}

// GetSystemStatus returns detailed system status
func (h *Handler) GetSystemStatus(c *gin.Context) {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)

	status := gin.H{
		"uptime":        time.Since(startTime).String(),
		"goroutines":    runtime.NumGoroutine(),
		"memory_alloc":  m.Alloc,
		"memory_sys":    m.Sys,
		"heap_alloc":    m.HeapAlloc,
		"heap_sys":      m.HeapSys,
		"heap_objects":  m.HeapObjects,
		"gc_cycles":     m.NumGC,
		"mihomo_status": "unknown",
	}

	// Get mihomo status
	if runtime, err := h.runtimeStore.Get(); err == nil {
		status["mihomo_status"] = runtime.Status
		status["mihomo_pid"] = runtime.PID
		status["mihomo_port"] = runtime.Port
		status["mihomo_uptime"] = runtime.Uptime
		status["mihomo_memory"] = runtime.MemoryUsage
	}

	if action, details, ts, ok := lastAutoUpdateSummary(h.auditStore); ok {
		status["last_auto_update"] = gin.H{
			"action":  action,
			"details": details,
			"at":      ts,
		}
	}

	if action, details, ts, ok := lastAlertSummary(h.auditStore); ok {
		status["last_alert"] = gin.H{
			"action":  action,
			"details": details,
			"at":      ts,
		}
	}

	h.respondJSON(c, http.StatusOK, status)
}

// GetSettings returns all system settings
func (h *Handler) GetSettings(c *gin.Context) {
	settings, err := h.settingStore.GetAll()
	if err != nil {
		h.respondError(c, http.StatusInternalServerError, "Failed to get settings")
		return
	}

	h.respondJSON(c, http.StatusOK, settings)
}

// UpdateSetting updates a system setting
func (h *Handler) UpdateSetting(c *gin.Context) {
	var req struct {
		Key         string `json:"key" binding:"required"`
		Value       string `json:"value" binding:"required"`
		Description string `json:"description"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		h.respondError(c, http.StatusBadRequest, "Invalid request: "+err.Error())
		return
	}

	if err := h.settingStore.Set(req.Key, req.Value, req.Description); err != nil {
		h.respondError(c, http.StatusInternalServerError, "Failed to update setting")
		return
	}

	h.auditLog(c, "update_setting", "setting", "Updated setting: "+req.Key)
	h.respondSuccess(c, "Setting updated successfully", nil)
}

// UpdateSettingsBatch updates multiple system settings
func (h *Handler) UpdateSettingsBatch(c *gin.Context) {
	var req struct {
		Settings []struct {
			Key         string `json:"key" binding:"required"`
			Value       string `json:"value" binding:"required"`
			Description string `json:"description"`
		} `json:"settings" binding:"required,min=1"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		h.respondError(c, http.StatusBadRequest, "Invalid request: "+err.Error())
		return
	}

	items := make([]types.Setting, 0, len(req.Settings))
	for _, s := range req.Settings {
		items = append(items, types.Setting{
			Key:         s.Key,
			Value:       s.Value,
			Description: s.Description,
		})
	}

	if err := h.settingStore.SetBatch(items); err != nil {
		h.respondError(c, http.StatusInternalServerError, "Failed to update settings")
		return
	}

	h.auditLog(c, "update_settings_batch", "setting", "Updated settings batch")
	h.respondSuccess(c, "Settings updated successfully", nil)
}

// GetAuditLogs returns audit logs
func (h *Handler) GetAuditLogs(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	limit := c.DefaultQuery("limit", "50")

	// Parse pagination
	pageInt := 1
	limitInt := 50
	if p, err := parseInt(page); err == nil && p > 0 {
		pageInt = p
	}
	if l, err := parseInt(limit); err == nil && l > 0 && l <= 100 {
		limitInt = l
	}

	offset := (pageInt - 1) * limitInt

	logs, err := h.auditStore.List(limitInt, offset)
	if err != nil {
		h.respondError(c, http.StatusInternalServerError, "Failed to get audit logs")
		return
	}

	total, err := h.auditStore.Count()
	if err != nil {
		total = 0
	}

	h.respondJSON(c, http.StatusOK, gin.H{
		"logs":  logs,
		"total": total,
		"page":  pageInt,
		"limit": limitInt,
	})
}

func parseInt(s string) (int, error) {
	var i int
	_, err := fmt.Sscanf(s, "%d", &i)
	return i, err
}

func hydrateLastAutoUpdateSummary(info *SystemInfo, auditStore interface {
	List(limit, offset int) ([]types.AuditLog, error)
}) {
	action, details, ts, ok := lastAutoUpdateSummary(auditStore)
	if !ok {
		return
	}
	info.LastAutoUpdateAction = action
	info.LastAutoUpdateDetails = details
	info.LastAutoUpdateAt = &ts
}

func hydrateLastAlertSummary(info *SystemInfo, auditStore interface {
	List(limit, offset int) ([]types.AuditLog, error)
}) {
	action, details, ts, ok := lastAlertSummary(auditStore)
	if !ok {
		return
	}
	info.LastAlertAction = action
	info.LastAlertDetails = details
	info.LastAlertAt = &ts
}

func lastAutoUpdateSummary(auditStore interface {
	List(limit, offset int) ([]types.AuditLog, error)
}) (action, details string, at time.Time, ok bool) {
	if auditStore == nil {
		return "", "", time.Time{}, false
	}

	logs, err := auditStore.List(100, 0)
	if err != nil {
		return "", "", time.Time{}, false
	}

	for _, log := range logs {
		if log.Resource != "mihomo" {
			continue
		}
		if !strings.HasPrefix(log.Action, "mihomo_update_") {
			continue
		}
		return log.Action, log.Details, log.CreatedAt, true
	}

	return "", "", time.Time{}, false
}

func lastAlertSummary(auditStore interface {
	List(limit, offset int) ([]types.AuditLog, error)
}) (action, details string, at time.Time, ok bool) {
	if auditStore == nil {
		return "", "", time.Time{}, false
	}

	logs, err := auditStore.List(200, 0)
	if err != nil {
		return "", "", time.Time{}, false
	}

	for _, log := range logs {
		if log.Resource != "alert" {
			continue
		}
		if !strings.HasPrefix(log.Action, "alert_") {
			continue
		}
		return log.Action, log.Details, log.CreatedAt, true
	}

	return "", "", time.Time{}, false
}

// GetNetworkInterfaces returns local IPv4 addresses for proxy host selection.
func (h *Handler) GetNetworkInterfaces(c *gin.Context) {
	addrs := []string{"127.0.0.1"}

	ifaces, err := net.Interfaces()
	if err == nil {
		for _, iface := range ifaces {
			if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
				continue
			}
			ifAddrs, err := iface.Addrs()
			if err != nil {
				continue
			}
			for _, a := range ifAddrs {
				var ip net.IP
				switch v := a.(type) {
				case *net.IPNet:
					ip = v.IP
				case *net.IPAddr:
					ip = v.IP
				}
				if ip == nil || ip.IsLoopback() || ip.To4() == nil {
					continue
				}
				addrs = append(addrs, ip.String())
			}
		}
	}

	h.respondJSON(c, http.StatusOK, gin.H{"addresses": addrs})
}

File: web-ui/src/stores/system.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as systemApi from '@/api/system'
import { useProxyStore } from '@/stores/proxy'

export type SystemInfo = systemApi.SystemInfo
export type SystemStatus = systemApi.SystemStatus

let ws: WebSocket | null = null
let reconnectTimer: number | null = null
let reconnectDelay = 1000

export const useSystemStore = defineStore('system', () => {
  const info = ref<SystemInfo | null>(null)
  const status = ref<SystemStatus | null>(null)
  const settings = ref<Record<string, string>>({})
  const loading = ref(false)
  const error = ref<string | null>(null)
  const wsConnected = ref(false)

  const fetchInfo = async (): Promise<void> => {
    loading.value = true
    error.value = null
    try {
      info.value = await systemApi.getSystemInfo()
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch system info'
      throw err
    } finally {
      loading.value = false
    }
  }

  const fetchStatus = async (): Promise<void> => {
    loading.value = true
    error.value = null
    try {
      status.value = await systemApi.getSystemStatus()
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch system status'
      throw err
    } finally {
      loading.value = false
    }
  }

  const fetchSettings = async (): Promise<void> => {
    loading.value = true
    error.value = null
    try {
      const data = await systemApi.getSettings()
      settings.value = data.reduce((acc, s) => {
        acc[s.key] = s.value
        return acc
      }, {} as Record<string, string>)
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch settings'
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateSetting = async (key: string, value: string, description?: string): Promise<void> => {
    loading.value = true
    error.value = null
    try {
      await systemApi.updateSetting(key, value, description)
      settings.value[key] = value
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to update setting'
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateSettingsBatch = async (newSettings: Record<string, string>): Promise<void> => {
    loading.value = true
    error.value = null
    try {
      await systemApi.updateSettingsBatch(
        Object.entries(newSettings).map(([key, value]) => ({ key, value: String(value) }))
      )
      settings.value = { ...settings.value, ...newSettings }
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to update settings'
      throw err
    } finally {
      loading.value = false
    }
  }

  const applySnapshot = (payload: any): void => {
    const statusPayload = payload?.status
    if (statusPayload) {
      status.value = {
        ...(status.value || {
          uptime: '', goroutines: 0, memory_alloc: 0, memory_sys: 0,
          heap_alloc: 0, heap_sys: 0, heap_objects: 0, gc_cycles: 0, mihomo_status: 'unknown'
        }),
        ...statusPayload
      }
      if (info.value) {
        info.value = {
          ...info.value,
          mihomo_status: String(statusPayload.mihomo_status || info.value.mihomo_status)
        }

        const wsAutoUpdate = statusPayload.last_auto_update
        if (wsAutoUpdate) {
          info.value.last_auto_update_action = wsAutoUpdate.action
          info.value.last_auto_update_details = wsAutoUpdate.details
          info.value.last_auto_update_at = wsAutoUpdate.at
        }

        const wsAlert = statusPayload.last_alert
        if (wsAlert) {
          info.value.last_alert_action = wsAlert.action
          info.value.last_alert_details = wsAlert.details
          info.value.last_alert_at = wsAlert.at
        }
      }
    }
  }

  const connectWS = (): void => {
    const token = localStorage.getItem('token')
    if (!token || ws) return

    const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
    const wsUrl = `${proto}//${window.location.host}/api/v1/system/ws?token=${encodeURIComponent(token)}`
    ws = new WebSocket(wsUrl)

    ws.onopen = () => {
      wsConnected.value = true
      reconnectDelay = 1000
      if (reconnectTimer) {
        window.clearTimeout(reconnectTimer)
        reconnectTimer = null
      }
    }

    ws.onmessage = (evt) => {
      try {
        const payload = JSON.parse(evt.data)
        if (payload?.type === 'snapshot') {
          applySnapshot(payload)
          if (payload.traffic) {
            const proxyStore = useProxyStore()
            proxyStore.applyTraffic(payload.traffic)
          }
        }
      } catch {
        // ignore invalid payload
      }
    }

    ws.onclose = () => {
      wsConnected.value = false
      ws = null
      reconnectTimer = window.setTimeout(() => {
        connectWS()
      }, reconnectDelay)
      reconnectDelay = Math.min(reconnectDelay * 2, 15000)
    }

    ws.onerror = () => {
      wsConnected.value = false
    }
  }

  const disconnectWS = (): void => {
    if (reconnectTimer) {
      window.clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
    if (ws) {
      ws.close()
      ws = null
    }
    wsConnected.value = false
  }

  return {
    info,
    status,
    settings,
    loading,
    error,
    wsConnected,
    fetchInfo,
    fetchStatus,
    fetchSettings,
    updateSetting,
    updateSettingsBatch,
    connectWS,
    disconnectWS
  }
})
